src/compiler/parser.ts

```diff
@@ -2461,6 +2464,22 @@ module ts {
 
             return result;
         }
+        
+        function parseTypePredicateOrHigher(): TypeNode {
+            let type = parseUnionTypeOrHigher();
+            if (token === SyntaxKind.IsKeyword &&
+                type.kind === SyntaxKind.TypeReference &&
+                (<TypeReferenceNode>type).typeName.kind === SyntaxKind.Identifier) {

```

Comment:
This approach is also not desirable because you end up having to extract out a child node of another node.  As mentioned before, you can change parseTypeReference and then make decision immediately after parsing teh entityName out.  It will make things simpler and push the check to the most immediate point possible.